doc/man/ossfs.1
.TH OSSFS "1" "February 2011" "OSSFS" "User Commands"
.SH NAME
OSSFS-Fuse \- FUSE-based file system backed by Aliyun OSS
.SH SYNOPSIS
.SS mounting
.TP
\fBossfs bucket[:/path] mountpoint \fP [options]
.SS unmounting
.TP
\fBumount mountpoint
.SS utility mode ( remove interrupted multipart uploading objects )
.TP
\fBossfs \-u bucket
.SH DESCRIPTION
ossfs is a FUSE filesystem that allows you to mount an Aliyun OSS bucket as a local filesystem. It stores files natively and transparently in OSS (i.e., you can use other programs to access the same files).
.SH AUTHENTICATION
The ossfs password file has this format (use this format if you have only one set of credentials):
.RS 4
\fBaccessKeyId\fP:\fBsecretAccessKey\fP
.RE

If you have more than one set of credentials, this syntax is also recognized:
.RS 4
\fBbucketName\fP:\fBaccessKeyId\fP:\fBsecretAccessKey\fP
.RE
.PP
Password files can be stored in two locations:
.RS 4
 \fB/etc/passwd-ossfs\fP     [0640]
 \fB$HOME/.passwd-ossfs\fP   [0600]
.RE
.SH OPTIONS
.SS "general options"
.TP
\fB\-h\fR   \fB\-\-help\fR
print help
.TP
\fB\  \fR   \fB\-\-version\fR
print version
.TP
\fB\-f\fR
FUSE foreground option - do not run as daemon.
.TP
\fB\-s\fR
FUSE singlethreaded option (disables multi-threaded operation)
.SS "mount options"
.TP
All ossfs options must given in the form where "opt" is:
 <option_name>=<option_value>
.TP
\fB\-o\fR default_acl (default="private")
the default canned acl to apply to all written OSS objects, e.g., "public-read".
Any created files will have this canned acl.
Any updated files will also have this canned acl applied!
.TP
\fB\-o\fR prefix (default="") (coming soon!)
a prefix to append to all OSS objects.
.TP
\fB\-o\fR retries (default="3")
number of times to retry a failed OSS transaction.
.TP
\fB\-o\fR use_cache (default="" which means disabled)
local folder to use for local file cache.
.TP
\fB\-o\fR del_cache - delete local file cache
delete local file cache when ossfs starts and exits.
.TP
\fB\-o\fR use_rrs (default is disable)
use Aliyun's Reduced Redundancy Storage.
this option can not be specified with use_sse.
(can specify use_rrs=1 for old version)
.TP
\fB\-o\fR use_sse (default is disable)
use Aliyun's Server Site Encryption.
this option can not be specified with use_rrs.
(can specify use_sse=1 for old version)
.TP
\fB\-o\fR passwd_file (default="")
specify the path to the password file, which which takes precedence over the password in $HOME/.passwd-ossfs and /etc/passwd-ossfs
.TP
\fB\-o\fR ahbe_conf (default="" which means disabled)
This option specifies the configuration file path which file is the additional HTTP header by file(object) extension.
 The configuration file format is below:
 -----------
 line         = [file suffix] HTTP-header [HTTP-values]
 file suffix  = file(object) suffix, if this field is empty, it means "*"(all object).
 HTTP-header  = additional HTTP header name
 HTTP-values  = additional HTTP header value
 -----------
 Sample:
 -----------
 .gz      Content-Encoding     gzip
 .Z       Content-Encoding     compress
          X-OSSFS-MYHTTPHEAD    myvalue
 -----------
 A sample configuration file is uploaded in "test" directory.
If you specify this option for set "Content-Encoding" HTTP header, please take care for RFC 2616.
.TP
\fB\-o\fR public_bucket (default="" which means disabled)
anonymously mount a public bucket when set to 1, ignores the $HOME/.passwd-ossfs and /etc/passwd-ossfs files.
.TP
\fB\-o\fR connect_timeout (default="300" seconds)
time to wait for connection before giving up.
.TP
\fB\-o\fR readwrite_timeout (default="60" seconds)
time to wait between read/write activity before giving up.
.TP
\fB\-o\fR max_stat_cache_size (default="10000" entries (about 4MB))
maximum number of entries in the stat cache
.TP
\fB\-o\fR stat_cache_expire (default is no expire)
specify expire time(seconds) for entries in the stat cache
.TP
\fB\-o\fR enable_noobj_cache (default is disable)
enable cache entries for the object which does not exist.
ossfs always has to check whether file(or sub directory) exists under object(path) when ossfs does some command, since ossfs has recognized a directory which does not exist and has files or sub directories under itself.
It increases ListBucket request and makes performance bad.
You can specify this option for performance, ossfs memorizes in stat cache that the object(file or directory) does not exist.
.TP
\fB\-o\fR nodnscache - disable dns cache.
ossfs is always using dns cache, this option make dns cache disable.
.TP
\fB\-o\fR nosscache - disable ssl session cache.
ossfs is always using ssl session cache, this option make ssl session cache disable.
.TP
\fB\-o\fR multireq_max (default="20")
maximum number of parallel request for listing objects.
.TP
\fB\-o\fR parallel_count (default="5")
number of parallel request for uploading big objects.
ossfs uploads large object(over 20MB) by multipart post request, and sends parallel requests.
This option limits parallel request count which ossfs requests at once.
It is necessary to set this value depending on a CPU and a network band.
.TP
\fB\-o\fR fd_page_size(default="52428800"(50MB))
number of internal management page size for each file descriptor.
For delayed reading and writing by ossfs, ossfs manages pages which is separated from object. Each pages has a status that data is already loaded(or not loaded yet).
This option should not be changed when you don't have a trouble with performance.
.TP
\fB\-o\fR url (default="http://s3.amazonaws.com")
sets the url to use to access Aliyun OSS. If you want to use HTTPS, then you can set url=https://s3.amazonaws.com
.TP
\fB\-o\fR nomultipart - disable multipart uploads
.TP
\fB\-o\fR enable_content_md5 ( default is disable )
verifying uploaded data without multipart by content-md5 header.
Enable to send "Content-MD5" header when uploading a object without multipart posting.
If this option is enabled, it has some influences on a performance of ossfs when uploading small object.
Because ossfs always checks MD5 when uploading large object, this option does not affect on large object.
.TP
\fB\-o\fR iam_role ( default is no role )
set the IAM Role that will supply the credentials from the instance meta-data.
.TP
\fB\-o\fR noxmlns - disable registing xml name space.
disable registing xml name space for response of ListBucketResult and ListVersionsResult etc. Default name space is looked up from "http://s3.amazonaws.com/doc/2006-03-01".
This option should not be specified now, because ossfs looks up xmlns automatically after v1.66.
.TP
\fB\-o\fR nocopyapi - for other incomplete compatibility object storage.
For a distributed object storage which is compatibility OSS API without PUT(copy api).
If you set this option, ossfs do not use PUT with "x-amz-copy-source"(copy api). Because traffic is increased 2-3 times by this option, we do not recommend this.
.TP
\fB\-o\fR norenameapi - for other incomplete compatibility object storage.
For a distributed object storage which is compatibility OSS API without PUT(copy api).
This option is a subset of nocopyapi option. The nocopyapi option does not use copy-api for all command(ex. chmod, chown, touch, mv, etc), but this option does not use copy-api for only rename command(ex. mv).
If this option is specified with nocopapi, the ossfs ignores it.
.SH FUSE/MOUNT OPTIONS
.TP
Most of the generic mount options described in 'man mount' are supported (ro, rw, suid, nosuid, dev, nodev, exec, noexec, atime, noatime, sync async, dirsync).  Filesystems are mounted with '-onodev,nosuid' by default, which can only be overridden by a privileged user.
.TP
There are many FUSE specific mount options that can be specified. e.g. allow_other. See the FUSE README.md for the full set.
.SH NOTES
.TP
Maximum file size=64GB (limited by ossfs, not Aliyun).
.TP
If enabled via the "use_cache" option, ossfs automatically maintains a local cache of files in the folder specified by use_cache. Whenever ossfs needs to read or write a file on OSS, it first downloads the entire file locally to the folder specified by use_cache and operates on it. When fuse_release() is called, ossfs will re-upload the file to OSS if it has been changed. ossfs uses md5 checksums to minimize downloads from OSS.
.TP
The folder specified by use_cache is just a local cache. It can be deleted at any time. ossfs rebuilds it on demand.
.TP
Local file caching works by calculating and comparing md5 checksums (ETag HTTP header).
.TP
ossfs leverages /etc/mime.types to "guess" the "correct" content-type based on file name extension. This means that you can copy a website to OSS and serve it up directly from OSS with correct content-types!
.SH BUGS
Due to OSS's "eventual consistency" limitations, file creation can and will occasionally fail. Even after a successful create, subsequent reads can fail for an indeterminate time, even after one or more successful reads. Create and read enough files and you will eventually encounter this failure. This is not a flaw in ossfs and it is not something a FUSE wrapper like ossfs can work around. The retries option does not address this issue. Your application must either tolerate or compensate for these failures, for example by retrying creates or reads.
.SH AUTHOR
ossfs has been written by Randy Rizun <rrizun@gmail.com>.


ChangeLog
ChangeLog for OSSFS
-------------------

Version 1.1 -- Mon Oct 18 2010

Dan Moore reopens the project and fixes various issues that had accumulated in the tracker. Adrian Petrescu converts the project to autotools and posts it to GitHub.

Version 1.0 -- 2008

Randy Rizun releases a basic version of S3FS on Google Code.


ossfs.py
#!/usr/bin/env python2.7
# -*- coding: utf-8 -*-

from re import search
from json import load
from argparse import ArgumentParser
from commands import getoutput, getstatusoutput

lsb_release = None

def configure():

    print u"现在准备编译环境"
    status, output = getstatusoutput("./autogen.sh")
    if status != 0:
        print u"准备编译环境失败，输出："
        print output
        exit(1)
    
    status, output = getstatusoutput("./configure")
    if status != 0:
        print u"准备编译环境失败，输出："
        print output
        exit(1)

def make():

    print u"现在重新编译 ./src/ossfs"
    status, output = getstatusoutput("make")
    if status != 0:
        print u"编译失败，输出："
        print output
        exit(1)

def start():

    # check binary file
    status, output = getstatusoutput("[ -f ./src/ossfs ]")
    if status != 0:
        print u"./src/ossfs 文件不存在，需要重新编译 ./src/ossfs"
        configure()
        make()

    # get ossfs.json
    with open("./ossfs.json", 'r') as fd:
        parameter_dict = load(fd)

    # check ossfs.json
    if parameter_dict["bucket_name"] in ["", None]:
        print u"ossfs.json 配置错误，bucket_name 字段为空"
        exit(1)
        
    if parameter_dict["access_id"] in ["", None]:
        print u"ossfs.json 配置错误，access_id 字段为空"
        exit(1)
        
    if parameter_dict["access_key"] in ["", None]:
        print u"ossfs.json 配置错误，access_key 字段为空"
        exit(1)
        
    if parameter_dict["mount_dir"] in ["", None]:
        print u"ossfs.json 配置错误，mount_dir 字段为空"
        exit(1)
        
    if parameter_dict["region_url"] in ["", None]:
        print u"ossfs.json 配置错误，url 字段为空"
        exit(1)

    # check mount_dir
    status, output = getstatusoutput("[ -d %s ]" % parameter_dict["mount_dir"])
    if status != 0:
        print u"mount_dir 该目录不存在，%s，ossfs 将递归创建该目录" % parameter_dict["mount_dir"]
        status, output = getstatusoutput("mkdir --parents %s" % parameter_dict["mount_dir"])
        if status != 0:
            print u"创建目录 %s 失败，输出：" % parameter_dict["mount_dir"]
            print output
            exit(1)

    status, output = getstatusoutput("df %s" % parameter_dict["mount_dir"])
    if status != 0:
        print u"检查目录失败，命令 df %s，输出：" % parameter_dict["mount_dir"]
        print output
        exit(1)
    if output.split()[7] == "ossfs":
        print u"检查目录失败，目录 %s 已经成为挂载点：" % parameter_dict["mount_dir"]
        print output
        exit(1)
            
    status, output = getstatusoutput("ls %s" % parameter_dict["mount_dir"])
    if status != 0:
        print u"检查目录失败，命令 ls %s，输出：" % output
        print output
        exit(1)
    if output != "":
        print u"检查目录失败，目录 %s 非空，内含：" % parameter_dict["mount_dir"]
        print output
        exit(1)

    # create $HOME/.passwd-ossfs
    status, output = getstatusoutput("echo '%s:%s:%s' > $HOME/.passwd-ossfs" % (parameter_dict["bucket_name"].strip(),
                                                                                parameter_dict["access_id"].strip(),
                                                                                parameter_dict["access_key"].strip()))
    if status != 0:
        print u"创建/覆盖文件 $HOME/.passwd-ossfs 失败，输出："
        print output
        exit(1)

    status, output = getstatusoutput("chmod 600 $HOME/.passwd-ossfs")
    if status != 0:
        print u"更改 $HOME/.passwd-ossfs 读写权限失败，输出："
        print output
        exit(1)

    # start
    status, output = getstatusoutput("./src/ossfs %s %s -o passwd_file=%s,url=%s" % (parameter_dict["bucket_name"].strip(),
                                                                                     parameter_dict["mount_dir"].strip(),
                                                                                     "$HOME/.passwd-ossfs",
                                                                                     parameter_dict["region_url"].strip()))
    if status != 0:
        print u"挂载失败，输出："
        print output
        exit(1)
    else:
        print u"启动成功，bucket %s 已经挂载在目录 %s 上" % (parameter_dict["bucket_name"],
                                                             parameter_dict["mount_dir"])

def stop():

    # get ossfs.json
    with open("./ossfs.json", 'r') as fd:
        parameter_dict = load(fd)

    # check ossfs.json
    if parameter_dict["bucket_name"] in ["", None]:
        print u"ossfs.json 配置错误，bucket_name 字段为空"
        exit(1)
        
    if parameter_dict["mount_dir"] in ["", None]:
        print u"ossfs.json 配置错误，mount_dir 字段为空"
        exit(1)

    running_flag = None
    status, output = getstatusoutput("ps -ef | grep './src/ossfs %s %s' | grep -v 'grep'" % (parameter_dict["bucket_name"],
                                                                                             parameter_dict["mount_dir"]))
    if status != 0:
        running_flag = False
    else:
        running_flag = True
    if running_flag:
        ossfs_pid = output.split()[2]
        status, output = getstatusoutput("kill -9 %s" % ossfs_pid)
        if status != 0:
            print u"停止失败，输出："
            print output
            exit(1)

    mounted_flag = None # whether mount_dir has been mounted
    status, output = getstatusoutput("df %s" % parameter_dict["mount_dir"])
    if status != 0:
        print u"检查目录失败，命令 df %s，输出：" % parameter_dict["mount_dir"]
        print output
        exit(1)
    if len(output.split()) != 13:
        print u"检查目录失败，命令 df %s，输出：" % parameter_dict["mount_dir"]
        print output
        exit(1)
    if output.split()[7] == "ossfs":
        mounted_flag = True # mounted, need to be umount
    else:
        mounted_flag = False # already umounted, needn't to be umount

    if mounted_flag is True:
        status, output = getstatusoutput("umount %s" % parameter_dict["mount_dir"])
        if status != 0:
            print u"卸载失败，输出："
            print output
            exit(1)

    print "停止并卸载成功"

def restart():

    stop()
    start()

def main():

    parser = ArgumentParser()
    parser.add_argument("command", help=u"ossfs 命令，可选项包括 start | stop | restart")
    
    args = parser.parse_args()

    # check release version
    status, output = getstatusoutput("lsb_release -a")
    if status != 0:
        print u"Linux发行版检查失败，输出："
        print output
        exit(1)
    lsb_release = None
    for i in range(0, len(output.split("\n"))):
        if search( "Description:", output.split("\n")[i]) is not None:
            lsb_release = output.split("\n")[i].strip()
            break
    if lsb_release is None:
        print u"Linux发行版检查失败，命令 lsb_release -a，输出："
        print output
        exit(1)
    if search("Ubuntu 14.04", lsb_release) is not None:
        lsb_release = "Ubuntu"
    elif search("CentOS Linux release 7", lsb_release) is not None:
        lsb_release = "CentOS"
    else:
        print u"Linux发行版检查失败，发行版必须是 Ubuntu 14.04 或者 CentOS 7，您当前的发行版是：", lsb_release
        exit(1)

    # check user
    status, output = getstatusoutput("echo ${USER}")
    if status != 0:
        print u"用户权限检查失败，输出："
        print output
        print exit(1)
    if output != "root":
        print u"用户必须具有root权限，您当前的用户是：%s", output
        print exit(1)

    # check command
    if args.command not in ["start", "stop", "restart"]:
        print u"命令必须是 start | stop | restart 其中之一, 你输入的是 '%s'" % args.command.decode("utf-8")
        exit(1)

    # check current dir
    status, output = getstatusoutput("ls")
    if "src" not in output.split():
        print u"你需要在ossfs项目的根目录进行执行 ossfs 的启停，当前所在目录并不是 ossfs 项目的根目录，当前目录是 '%s'" % getoutput("pwd").decode("utf-8")
        exit(1)
    if "ossfs.py" not in output.split():
        print u"你需要在ossfs项目的根目录进行执行 ossfs 的启停，当前所在目录并不是 ossfs 项目的根目录，当前目录是 '%s'" % getoutput("pwd").decode("utf-8")
        exit(1)
    if "ossfs.json" not in output.split():
        print u"你需要在ossfs项目的根目录进行执行 ossfs 的启停，当前所在目录并不是 ossfs 项目的根目录，当前目录是 '%s'" % getoutput("pwd").decode("utf-8")
        exit(1)
        
    function_dict = {
        "start" : start,
        "stop" : stop,
        "restart" : restart
    }
    function_dict[args.command]()
    
if __name__ == "__main__":

    main()
